File: Aliyev-Vladislav/Semester_1/Block_3/Task_3(filters)/filters.c
#include <math.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include "filters.h"
#include "tools.h"

void gray(struct pixel* pixels, int height, int width)
{
	int total;
	for (int i = 0; i < height * width; i++)
	{
		total = (pixels[i].r + pixels[i].g + pixels[i].b) / 3;
		pixels[i].r = total;
		pixels[i].g = total;
		pixels[i].b = total;
	}
}

float gauss(int* a)
{
	float total = 0;
	for (int i = 0; i < 9; i++)
	{
		if ((i == 0) || (i == 2) || (i == 6) || (i == 8))
			total += a[i] / 16;
		if ((i == 1) || (i == 3) || (i == 5) || (i == 7))
			total += a[i] / 8;
		if (i == 4)
			total += a[i] / 4;
	}
	return total;
}

int sobelX(int* a)
{
	float total = 0;
	for (int i = 0; i < 9; i++)
	{
		if ((i == 0) || (i == 6))
			total -= a[i];
		if (i == 3)
			total -= a[i] * 2;
		if ((i == 2) || (i == 8))
			total += a[i];
		if (i == 5)
			total += a[i] * 2;
	}
	return (int)min(255, max(0, abs(total)));
}

int sobelY(int* a)
{
	float total = 0;
	for (int i = 0; i < 9; i++)
	{
		if ((i == 0) || (i == 2))
			total -= a[i];
		if (i == 1)
			total -= a[i] * 2;
		if ((i == 6) || (i == 8))
			total += a[i];
		if (i == 7)
			total += a[i] * 2;
	}
	return (int)min(255, max(0, abs(total)));
}

void filter(struct pixel* pixels, int height, int width, char* filter)
{
	struct pixel* pixels_copy;
	pixels_copy = (struct pixel*)malloc(width * height * sizeof(struct pixel));
	memcpy(pixels_copy, pixels, width * height * sizeof(struct pixel));
	int* red;
	int* green;
	int* blue;
	red = (int*)malloc(9 * sizeof(int));
	green = (int*)malloc(9 * sizeof(int));
	blue = (int*)malloc(9 * sizeof(int));
	for (int i = 0; i < height * width - 2 * width; i++)
	{
		for (int j = 0; j < 3; j++)
		{
			for (int k = 0; k < 3; k++)
			{
				if ((i % width != 0) && (i < (width * height - width)))
				{
					red[j + 3 * k] = pixels_copy[i + j + width * k].r;
					green[j + 3 * k] = pixels_copy[i + j + width * k].g;
					blue[j + 3 * k] = pixels_copy[i + j + width * k].b;
				}
			}
		}
		if (strcmp(filter, "median") == 0)
		{
			pixels[i + 1 + width].r = sort(red);
			pixels[i + 1 + width].g = sort(green);
			pixels[i + 1 + width].b = sort(blue);
		}
		if (strcmp(filter, "gauss") == 0)
		{
			pixels[i + 1 + width].r = (int)gauss(red);
			pixels[i + 1 + width].g = (int)gauss(green);
			pixels[i + 1 + width].b = (int)gauss(blue);
		}
		if (strcmp(filter, "sobelX") == 0)
		{
			pixels[i + 1 + width].r = (int)sobelX(red);
			pixels[i + 1 + width].g = (int)sobelX(green);
			pixels[i + 1 + width].b = (int)sobelX(blue);
		}
		if (strcmp(filter, "sobelY") == 0)
		{
			pixels[i + 1 + width].r = (int)sobelY(red);
			pixels[i + 1 + width].g = (int)sobelY(green);
			pixels[i + 1 + width].b = (int)sobelY(blue);
		}
	}
	free(red);
	free(green);
	free(blue);
	free(pixels_copy);
}

File: Zabelin Alexandr/Term 2/Block 1/Task 1/BitMapTask/Filters.cs
using System;

namespace BitMapTask
{
    static public class Filters
    {
        static private void ApplyingMatrix(BitMapImage image, float[,] matrix)
        {
            float r, g, b;
            Pixel[,] filteredImage = new Pixel[image.Width, image.Height];

            for (uint x = 1; x < image.Width - 1; x++)
            {
                for (uint y = 1; y < image.Height - 1; y++)
                {
                    r = 0;
                    g = 0;
                    b = 0;

                    for (uint row = 0; row < 3; row++)
                    {
                        for (uint col = 0; col < 3; col++)
                        {
                            r += image.GetPixel(x + col - 1, y + row - 1).Red * matrix[row, col];
                            g += image.GetPixel(x + col - 1, y + row - 1).Green * matrix[row, col];
                            b += image.GetPixel(x + col - 1, y + row - 1).Blue * matrix[row, col];
                        }
                    }

                    Math.Abs(r);
                    Math.Abs(g);
                    Math.Abs(b);
                    filteredImage[x, y] = new Pixel((byte)r, (byte)g, (byte)b);
                }
            }

            RewriteImage(image, filteredImage);
        }

        static private void RewriteImage(BitMapImage image, Pixel[,] filteredImage)
        {
            for (uint x = 1; x < image.Width - 1; x++)
            {
                for (uint y = 1; y < image.Height - 1; y++)
                {
                    image.SetPixel(x, y, filteredImage[x, y]);
                }
            }
        }

        static public void Median(BitMapImage image)
        {
            int[] r = new int[9], g = new int[9], b = new int[9];
            Pixel[,] filteredImage = new Pixel[image.Width, image.Height];

            for (uint x = 1; x < image.Width - 1; x++)
            {
                for (uint y = 1; y < image.Height - 1; y++)
                {
                    for (uint row = 0; row < 3; row++)
                    {
                        for (uint col = 0; col < 3; col++)
                        {
                            r[3 * row + col] = image.GetPixel(x + col - 1, y + row - 1).Red;
                            g[3 * row + col] = image.GetPixel(x + col - 1, y + row - 1).Green;
                            b[3 * row + col] = image.GetPixel(x + col - 1, y + row - 1).Blue;
                        }
                    }

                    Array.Sort(r);
                    Array.Sort(g);
                    Array.Sort(b);
                    filteredImage[x, y] = new Pixel((byte)r[4], (byte)g[4], (byte)b[4]);
                }
            }

            RewriteImage(image, filteredImage);
        }

        static public void Gauss(BitMapImage image)
        {
            float[,] matrix = new float[3, 3] {
                {1, 2, 1},
                {2, 4, 2},
                {1, 2, 1}
            };
            // Transforming float matrix

            for (int i = 0; i < 3; i++)
            {
                for (int j = 0; j < 3; j++)
                {
                    matrix[i, j] = matrix[i, j] / 16.0f;
                }
            }

            // -------------------------

            ApplyingMatrix(image, matrix);
        }

        static public void SobelX(BitMapImage image)
        {
            float[,] matrix = new float[3, 3] {
                {1, 0, -1},
                {2, 0, -2},
                {1, 0, -1}
            };

            ApplyingMatrix(image, matrix);
        }

        static public void SobelY(BitMapImage image)
        {
            float[,] matrix = new float[3, 3] {
                {1, 2, 1},
                {0, 0, 0},
                {-1, -2, -1}
            };

            ApplyingMatrix(image, matrix);
        }

        static public void Gray(BitMapImage image)
        {
            int mean;
            Pixel pixel;

            for (uint x = 0; x < image.Width; x++)
            {
                for (uint y = 0; y < image.Height; y++)
                {
                    pixel = image.GetPixel(x, y);
                    mean = (pixel.Red + pixel.Green + pixel.Blue) / 3;

                    pixel.Red = (byte)mean;
                    pixel.Green = (byte)mean;
                    pixel.Blue = (byte)mean;

                    image.SetPixel(x, y, pixel);
                }
            }
        }
    }
}